create_website.py

```diff
@@ -275,232 +312,138 @@ def create_plot(ds, all_data, xn, yn, linestyle, additional_label = "", plottype
+                    if get_dataset_from_desc(e) == ds and \
+                       get_distance_from_desc(e) == dm]
+            sorted_matches = sorted(matching_datasets, \
+                    key = lambda e: int(get_count_from_desc(e)))
+            for idd in sorted_matches:
+                output_str += get_row_desc(idd, get_dataset_label(idd))
     output_str += """
         <h2 id="algorithms">Results by Algorithm</h2>
         <ul class="list-inline"><b>Algorithms:</b>"""
-    algorithms = all_runs_by_algorithm.keys()
-    for algo in algorithms:
+    algorithm_names = algorithms.keys()
+    for algo in algorithm_names:
         output_str += '<li><a href="#%(name)s">%(name)s</a></li>' % {"name" : algo}
     output_str += "</ul>"
-    for algo in algorithms:
-        output_str += """
-            <a href="./%(name)s.html">
-            <div class="row" id="%(name)s">
-                <div class = "col-md-4 bg-success">
-                    <h4>%(name)s</h4>
-                    <dl class="dl-horizontal">
-                    """ % { "name" : algo }
-        if "alogs" in definitions and algo in definitions["algos"]:
-            for k in definitions["algos"][algo]:
-                output_str += """
-                        <dt>%s</dt>
-                        <dd>%s</dd>
-                        """ % (k, str(definitions["algos"][algo][k]))
-        output_str += """
-                </dl>
-            </div>
-            <div class = "col-md-8">
-                <img class = "img-responsive" src="%(name)s.png" />
-            </div>
-        </div>
-        </a>
-        <hr />""" % { "name" : algo}
-    output_str += """
-            <h2 id ="raw">Raw Data & Configuration</h2>
-            <p>Please find the raw experimental data <a href="./results-sisap.tar">here</a> (13 GB). The query set is available <a href="./queries-sisap.tar">queries-sisap.tar</a> (8 GB) as well. The algorithms used the following parameter choices in the experiments: <a href="./algos.yaml">k = 10</a> and <a href="./algos100.yaml">k=100</a>.</p>
-            <div id="contact">
-            <h2>Contact</h2>
-            <p>ANN-Benchmarks has been developed by Martin Aumueller (maau@itu.dk), Erik Bernhardsson (mail@erikbern.com), and Alec Faitfull (alef@itu.dk). Please use
-            <a href="https://github.com/maau/ann-benchmarks/">Github</a> to submit your implementation or improvements.</p>
-            </div>
-        </div>
-    </body>
-</html>"""
-    text_file.write(output_str)
+    for algo in algorithm_names:
+        output_str += get_row_desc(algo, algo)
+    output_str += get_index_footer()
+
+    with open(args.outputdir + "index.html", "w") as text_file:
+        text_file.write(output_str)
+
+def load_all_results():
+    """Read all result files and compute all metrics"""
+    all_runs_by_dataset = {}
+    all_runs_by_algorithm = {}
+    for f in results.load_all_results():
+        properties = dict(f.attrs)
```

Comment:
>> One problem a ran into:

>> For some algorithms, reading the stored files results in strings being represented as bytes (e.g., b'faiss-ivf'). I currently included a workaround for this in create_website.py. Do you know what is going on? (Running annoy and faiss-ivf on random-xs-20-angular results in this problem for faiss, but not for annoy on my machine.)

> I'm fairly sure this is because some algorithms run in Python 2, some in Python 3. I noticed some inconsistencies with bytes/strings too but it didn't have an impact on the benchmarks so I just ignored it... shouldn't be very hard to fix if it's causing issues

That should be it. Faiss still runs on Python 2. I checked some h5py resources, but couldn't find a solution. So, I'm converting the `f.attrs` dictionary myself. (I'm building strings from these descriptions, but I cannot do it when they are bytes.)